src/actions/sources/newSources.js

```diff
@@ -50,14 +49,16 @@ function loadSourceMaps(sources) {
       return;
     }
 
-    let originalSources = await Promise.all(
-      sources.map(({ id }) => dispatch(loadSourceMap(id)))
+    await Promise.all(
+      sources.map(({ id }) =>
+        dispatch(loadSourceMap(id)).then(originalSource => {
+          if (originalSource) {
+            dispatch(newSources(originalSource));
+          }
+          return originalSource;
+        })
+      )
```

Comment:
I don't think we need to return anything, so if do this it could be much simpler

```js
sources.map(async ({ id }) => {
  const originalSources = await dispatch(loadSourceMap(id));
   if (originalSources) {
      dispatch(newSources(originalSource));
    }
 })
```

My concern with this approach is that in the case of an angular app this could be 500+ redux calls to newSources..  Of course, we can add some throttling in `newSources` so at most we dispatch the new sources every 50ms.

Reply:
I was trying to think what a throttled addSources would look like and I think this is a reasonable approximation

```diff
diff --git a/src/actions/sources/newSources.js b/src/actions/sources/newSources.js
index 2c5b52d98..5466039e0 100644
--- a/src/actions/sources/newSources.js
+++ b/src/actions/sources/newSources.js
@@ -174,6 +174,14 @@ export function newSource(source: Source) {
   };
 }

+let throttledSources = []
+const addSources = debounce(() => {
+  return ({ dispatch, getState }: ThunkArgs) => {
+    dispatch(({ type: "ADD_SOURCES", sources: throttledSources }: Action));
+    throttledSources = [];
+  }
+}
+
 export function newSources(sources: Source[]) {
   return async ({ dispatch, getState }: ThunkArgs) => {
     sources = sources.filter(source => !getSource(getState(), source.id));
@@ -182,7 +190,8 @@ export function newSources(sources: Source[]) {
       return;
     }

-    dispatch(({ type: "ADD_SOURCES", sources: sources }: Action));
+    throttledSources = [...throttledSources, ...sources]
+    dispatch(addSources())

     await dispatch(loadSourceMaps(sources));
```

Reply:
Maybe an even smarter way to do this would be to re-use the same throttled source queue we use in [firefox/events.js](https://github.com/devtools-html/debugger.html/blob/master/src/client/firefox/events.js#L102-L103)

https://github.com/devtools-html/debugger.html/blob/master/src/utils/source-queue.js#L20


```diff
diff --git a/src/actions/sources/newSources.js b/src/actions/sources/newSources.js
index 2c5b52d98..3dd0d1e3f 100644
--- a/src/actions/sources/newSources.js
+++ b/src/actions/sources/newSources.js
@@ -25,6 +25,7 @@ import {
   getPendingBreakpointsForSource
 } from "../../selectors";
 
+import sourceQueue from "../../utils/source-queue";
 import type { Source, SourceId } from "../../types";
 import type { Action, ThunkArgs } from "../types";
 
@@ -50,14 +51,10 @@ function loadSourceMaps(sources) {
       return;
     }
 
-    let originalSources = await Promise.all(
-      sources.map(({ id }) => dispatch(loadSourceMap(id)))
-    );
-
-    originalSources = flatten(originalSources).filter(Boolean);
-    if (originalSources.length > 0) {
-      await dispatch(newSources(originalSources));
-    }
+    sources.map(async ({ id }) => {
+      const originalSources = await dispatch(loadSourceMap(id));
+      queueSources(originalSource);
+    });
   };
 }
 
diff --git a/src/utils/source-queue.js b/src/utils/source-queue.js
index 4c6428df8..0338edb8b 100644
--- a/src/utils/source-queue.js
+++ b/src/utils/source-queue.js
@@ -32,6 +32,11 @@ export default {
     queuedSources.push(source);
     queue();
   },
+  queueSources: sources => {
+    queuedSources = queuedSources.concat(sources);
+    queue();
+  },
+
   flush: () => Promise.all([queue.flush(), currentWork]),
   clear: () => queue.cancel()
};
```

This solution strikes me as particularly elegant as we're re-using the throttling that already exists and would be able to add new original sources pretty smoothly

Reply:
Alright, I've made newSources use source-queue and updated the unit tests accordingly. Mochitests are failing on travis now, going to try to reproduce those failures locally now.

Reply:
Running mochitests on my laptop takes a while so I've only managed to do 3 complete runs, the result is this:

```
Overall Summary
===============

mochitest-browser
~~~~~~~~~~~~~~~~~
Ran 2878 checks (73 tests, 2805 subtests)
Expected results: 2872
Skipped: 3 tests
Unexpected results: 2
  subtest: 2 (2 fail)

Unexpected Results
------------------
devtools/client/debugger/new/test/mochitest/browser_dbg-breakpoints-columns.js
  FAIL Test timed out -
devtools/client/debugger/new/test/mochitest/browser_dbg-tabs-keyboard.js
  FAIL Uncaught exception - at chrome://mochitests/content/browser/devtools/client/debugger/new/test/mochitest/helpers.js:605 - TypeError: findElement(...) is null; can't access its "children" property
Stack trace:
countTabs@chrome://mochitests/content/browser/devtools/client/debugger/new/test/mochitest/helpers.js:605:10
@chrome://mochitests/content/browser/devtools/client/debugger/new/test/mochitest/browser_dbg-tabs-keyboard.js:15:6
Async*Tester_execTest/<@chrome://mochikit/content/browser-test.js:1093:34
async*Tester_execTest@chrome://mochikit/content/browser-test.js:1084:16
nextTest/<@chrome://mochikit/content/browser-test.js:982:9
SimpleTest.waitForFocus/waitForFocusInner/focusedOrLoaded/<@chrome://mochikit/content/tests/SimpleTest/SimpleTest.js:803:59
```

Those two tests fail reproducibly. I'm not seeing all those time outs that seem to happen on travis.